Add `ArrowToParquetSchemaConverter`, deprecate `arrow_to_parquet_schema`

alamb opened the pull request:
# Which issue does this PR close?

- Part of https://github.com/apache/arrow-rs/issues/1938

# Rationale for this change
 
While working on the upgrade to arrow 54 in DataFusion https://github.com/apache/datafusion/pull/13663   I found it was not clear what the `coerce_types` argument in `arrow_to_parquet_schema` meant or what value to pick to keep the existing behavior. 

I had to go digging on https://github.com/apache/arrow-rs/pull/6313 to find what the default / old behavior was (`true` or `false`)

Turns out the answer is
```rust
    let schema_desc = arrow_to_parquet_schema(&schema, false)?;
```

It is clearly documented on `WriterProperties` but not on the lower level apis

Thus I would like to propose making a new API for converting schemas where the defaults are clearer and deprecate `arrow_to_parquet_schema` (and revert the unnecessary breaking API change)

# What changes are included in this PR?

1. Add `ArrowToParquetSchemaConverter`, 
2. Lots of comments / examples
3. Revert the changes to `arrow_to_parquet_schema` made in https://github.com/apache/arrow-rs/pull/6313
4. Deprecate `arrow_to_parquet_schema` with a clear message about what to use instread



# Are there any user-facing changes?
A new API and a deprecation of an existing one 

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please call them out.
-->

alamb commented on parquet/src/arrow/schema/mod.rs:
```diff
@@ -225,29 +225,99 @@ pub(crate) fn add_encoded_arrow_schema_to_metadata(schema: &Schema, props: &mut
-    arrow_to_parquet_schema_with_root(schema, "arrow_schema", coerce_types)
-}
+#[deprecated(since = "54.0.0", note = "Use `ArrowToParquetSchemaConverter` instead")]
+pub fn arrow_to_parquet_schema(schema: &Schema) -> Result<SchemaDescriptor> {
```
This reverts the changes made in https://github.com/apache/arrow-rs/pull/6313/files#diff-6a684124e78f254fa6ea23b98f04be0a3ff5a2e4b24890d9c6f4dd850ba11333L232

alamb commented on parquet/src/arrow/schema/mod.rs:
```diff
@@ -225,29 +225,99 @@ pub(crate) fn add_encoded_arrow_schema_to_metadata(schema: &Schema, props: &mut
+pub fn arrow_to_parquet_schema(schema: &Schema) -> Result<SchemaDescriptor> {
 
-/// Convert arrow schema to parquet schema specifying the name of the root schema element
-pub fn arrow_to_parquet_schema_with_root(
```
It turns out this function is not actually exported (it is pub in this module, but not pub exported):  

https://docs.rs/parquet/latest/parquet/?search=arrow_to_parquet_schema_with_root

Returns no results

The compiler told me it was unused once I switched everything over to use `ArrowToParquetSchemaConverter`

alamb commented on parquet/src/file/properties.rs:
```diff
@@ -287,15 +286,13 @@ impl WriterProperties {
+    ///
+    /// Setting this option to `true` will result in parquet files that can be
+    /// read by more readers, but may lose precision for arrow types such as
+    /// [`DataType::Date64`] which have no direct corresponding Parquet type.
```
I also felt it would help if we described more explicitly what the impact of enabling this option was (and left a link to the longer backstory)

etseidl commented:
@alamb could you take a look at #6828? Some of the doc changes will conflict, at least. +1 on the idea, though 😄

etseidl reviewed (commented):
Big improvement, thanks!

alamb commented on parquet/src/arrow/schema/mod.rs:
```diff
@@ -225,27 +225,130 @@ pub(crate) fn add_encoded_arrow_schema_to_metadata(schema: &Schema, props: &mut
 }
 
-/// Convert arrow schema to parquet schema
+/// Converter for arrow schema to parquet schema
```
I think we should strive for always capitalizing them `Arrow` and `Parquet` -- i will fix

alamb commented on parquet/src/arrow/schema/mod.rs:
```diff
@@ -225,27 +225,130 @@ pub(crate) fn add_encoded_arrow_schema_to_metadata(schema: &Schema, props: &mut
+    /// For example, some Date64 values may be truncated when stored with
+    /// parquet's native 32 bit date type. For [`List`] and [`Map`] types, some
+    /// Parquet readers expect certain schema elements to have specific names
+    /// (earlier versions of the spec was somewhat ambiguous on this point).
```
I think it is an improvment

alamb commented on parquet/src/file/properties.rs:
```diff
@@ -780,22 +779,16 @@ impl WriterPropertiesBuilder {
-    /// [`Map`]: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps
+    /// Setting this option to `true` will result in parquet files that can be
+    /// read by more readers, but may lose precision for arrow types such as
+    /// [`DataType::Date64`] which have no direct corresponding Parquet type.
```
I agree - I pushed e7b7d20eb3 to try and clarify -- let me know what you think

etseidl replied:
💯

alamb reviewed (commented):
Thank you for the review @etseidl

etseidl reviewed (approved):
Just the doc CI failure to fix. Otherwise looks ready to go (mod capitalization which can be a follow on 😉). Thanks again @alamb.

alamb commented:
> Just the doc CI failure to fix. Otherwise looks ready to go (mod capitalization which can be a follow on 😉). Thanks again @alamb.

Thank you 🙏 

I fixed a few more capitalization inconsistencies and added a link to the date section of the parquet spec too

tustvold commented on parquet/src/arrow/schema/mod.rs:
```diff
@@ -225,27 +219,134 @@ pub(crate) fn add_encoded_arrow_schema_to_metadata(schema: &Schema, props: &mut
+    }
+
+    /// Set the root schema element name (defaults to `"arrow_schema"`).
+    pub fn schema_root(mut self, schema_root: &'a str) -> Self {
```
```suggestion
    pub fn with_root(mut self, schema_root: &'a str) -> Self {
```
Perhaps? For consistency and because the schema is perhaps redundant?

alamb replied:
Actually, I think `with_schema_root` is more consistent with other options:

https://github.com/apache/arrow-rs/blob/9ffa06543be51613ea1f509e63f6e7405b7d9989/parquet/src/arrow/arrow_writer/mod.rs#L393-L392

tustvold commented on parquet/src/arrow/schema/mod.rs:
````diff
@@ -225,27 +219,134 @@ pub(crate) fn add_encoded_arrow_schema_to_metadata(schema: &Schema, props: &mut
+/// assert_eq!(parquet_schema, expected_parquet_schema);
+/// ```
+#[derive(Debug)]
+pub struct ArrowToParquetSchemaConverter<'a> {
````
```suggestion
pub struct ArrowSchemaConverter<'a> {
```
Perhaps, I don't feel strongly but might reduce the word salad

alamb replied:
It is also more consistent with things like https://docs.rs/parquet/latest/parquet/arrow/arrow_reader/struct.ArrowReaderBuilder.html -- I will change

tustvold reviewed (commented):
I did wonder if this should instead be an options struct that gets passed to a method (or has a member function that accepts the actual schema). This would allow using the same config for multiple files, and is perhaps more consistent with other APIs?

alamb commented:
> I did wonder if this should instead be an options struct that gets passed to a method (or has a member function that accepts the actual schema). This would allow using the same config for multiple files, and is perhaps more consistent with other APIs?

I agree, as there is no state held in the converter there is no reason for it to be a consuming builder. I made this change in 0ab4eb4ccc

alamb commented:
Thanks again @etseidl  and @tustvold
